Display a document icon when using the site editor

dsas opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Displaying an icon next to the title when using the site editor. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Many users are confused when entering the Site Editor - not understanding the context that they are editing elements under (within templates). This often results in content which is intended for a single page being added as content to the template (amongst other things). The Site Editor header displays the current document's name, but not its type. This change gives an additional hint to users what is being edited.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The `useEditedEntityRecord` hook now returns an icon for the entity being edited.
The `DocumentActions` component displays this icon in the document action title

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
 1. Open the site editor and edit a template - you should now see a template icon
 3. Open the site editor and edit a header template part - you should now see a header icon
 4. Open the site editor and edit a template part that isn't a header or footer - you should now see a template-part icon

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Keyboard navigation is unaffected.

## Screenshots or screencast <!-- if applicable -->



Title | Before | After
------|--------|------
 Editing template | <img width="1468" alt="Screenshot 2023-02-13 at 18 57 39" src="https://user-images.githubusercontent.com/93301/218588662-ec197250-0ebb-4ffe-b527-09bc34a8338a.png"> | <img width="1336" alt="image" src="https://user-images.githubusercontent.com/93301/233079662-c4d241db-a920-4423-87b2-c08b6c98e047.png">
 Editing header | <img width="1468" alt="Screenshot 2023-02-13 at 18 58 29" src="https://user-images.githubusercontent.com/93301/218588773-135d6e59-9c66-49b4-a4c9-bf0e7e0198d8.png"> | <img width="1336" alt="image" src="https://user-images.githubusercontent.com/93301/233079591-da70aa65-5755-4e3d-8f89-57d750b90379.png">
 Editing header as part of a template | <img width="1468" alt="Screenshot 2023-02-13 at 18 57 52" src="https://user-images.githubusercontent.com/93301/218588900-52392622-42d3-48fe-ab36-e3cb1216e0da.png"> | <img width="1336" alt="image" src="https://user-images.githubusercontent.com/93301/233079486-8ddcd382-54bd-471f-9bee-3e06a5203ec0.png">
 Editing template part | <img width="1468" alt="Screenshot 2023-02-13 at 18 58 40" src="https://user-images.githubusercontent.com/93301/218588962-3e782893-21d0-4177-bd4f-896710972c71.png"> | <img width="1336" alt="image" src="https://user-images.githubusercontent.com/93301/233079318-6a4a2442-eb63-4778-940e-deb4c0011a01.png">

dsas committed:
Return icon from useEditedEntityRecord

Returning the entity icon from useEditedEntityRecord so that it can be
displayed in related places in the UI.

dsas committed:
Site-editor: Display icon for entity being edited

Prefixing the template / template-part name with the icon of the entity
being edited provides an additional hint to the user as to what they're
editing.

jameskoster commented:
I think something like this can help, but we might need a more nuanced solution, or at least some more design exploration to ensure this is the correct direction. 

The truth is that while working on a single document in the site editor you can potentially end up editing several additional elements all at the same time. For example if you open the `category-recipes` template you could directly edit:

* Multiple posts and their associated meta data
* The 'Recipes' category data (title, description)
* Template Parts
* Reusable Blocks
* Site data (site title, logo, etc)
* Menus

We also need to think about how users interpret the 'document' in certain flows. For example if you navigate to Site Editor > Navigation > About > Edit – I think most people would feel like they're editing their About page primarily, rather than the `page` template.

All that considered, framing the session around a single document may be too blunt of an instrument. I wonder if there's a way we can make the title more dynamic so that it communicates the context for the selected block. We've explored this a little for template parts, can we take it further?

dsas commented:
@jameskoster, sorry can I just get some more clarity on the next steps here?

I'm not sure that this PR changes the level of single document framing. This PR prefixes the top-level document name with an appropriate icon. 

When editing a template-part while editing a template, the document name is currently already suffixed with the template-part icon and the name of the template part being edited. This is unchanged. 

Are you saying you want to see this template-part behaviour (or something similar) for entities other than template parts as part of this PR? 

What is a design exploration and how can I get one?

jameskoster commented:
Apologies I may have gone into the weeds a bit there. 

Would you mind rebasing and increasing the gap a bit between icon + title? Then we can get more reviews.

dsas committed:
Tweak document title spacing

 * Increase spacing between document/secondary titles and their icons by 4px
 * Increase spacing between document and secondary titles by 6px

dsas commented:
Cheers @jameskoster I've rebased and added spacing between a title and it's icon, and increased spacing between primary and secondary titles too to suit.

Getting the correct amounts of spacing between things is something I don't have much confidence in, so I'd love to hear some feedback. I've updated the after screenshots on the description so you can easily see what I've done.

jameskoster commented:
Thanks for updating :)

Rather than using CSS you might be interested in the [HStack component](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-hstack--default).

If you stick with css, let's use the `$grid-unit` variable for the values.

I think we can get broader feedback from @WordPress/gutenberg-design now.

dsas committed:
Use grid units rather than pixel values

dsas commented:
Thanks for the help @jameskoster, I've used $grid-unit variables for values now.

No objection to moving to HStack in general, but perhaps that should happen to the DocumentActions component regardless of whether this PR lands?

jasmussen commented:
This feels like a fine enough solution to me. 

Template icons:
<img width="216" alt="Screenshot 2023-04-20 at 09 52 22" src="https://user-images.githubusercontent.com/1204802/233298190-9dee780b-db00-43c4-ad2a-27b3ff083ea0.png">
<img width="161" alt="Screenshot 2023-04-20 at 09 52 40" src="https://user-images.githubusercontent.com/1204802/233298217-498bc6ef-d2b4-4203-b2fc-70a0acf87936.png">

Btw should either or both of those use the Home icon? Not a blocker for this PR, in fact probably a rabbit hole. But mentioning it nevertheless.

Header template part:

<img width="170" alt="Screenshot 2023-04-20 at 09 52 52" src="https://user-images.githubusercontent.com/1204802/233298337-12e7cd1c-3395-49e9-8d21-ed84dfbc6413.png">

Before we move too much further, I'd love a quick take by @SaxonF, as I know he's explored a different format for this top/center chip.

jameskoster commented:
> Btw should either or both of those use the Home icon? 

It's a good question. To connect the dot, we currently use icons in the Add template menu:

<img width="290" alt="Screenshot 2023-04-20 at 11 26 06" src="https://user-images.githubusercontent.com/846565/233339133-ec540484-7dfa-4065-b2af-6a082839b258.png">

It might be confusing to see different icons in these two locations, especially immediately after creating the template.

The icons in the add template menu are not canonical, and imo do not add much value. The most important distinction to make going forwards is whether you're editing content or a template, and in that respect fewer icons is probably better.

So with that said, I think the behavior in this PR is good, and as a follow-up we can remove the decorative icons in the add template menu.

paaljoachim commented:
**Brainstorming/rambling coming up.**

It seems like a very good idea to include the icon beside the template name.

**A before:**
![Screenshot 2023-04-21 at 11 31 46](https://user-images.githubusercontent.com/5323259/233601568-f4f296bc-9651-4d94-a47d-908bcd0e20b4.jpg)

**After:**
![Screenshot 2023-04-21 at 11 32 04](https://user-images.githubusercontent.com/5323259/233601607-f568119a-2b58-4f38-92c6-4f91f36cd9db.jpg)

Gradually from going from a yes this is a very good idea I began to become hesitant.
The sidebar icon on the right can be clicked to show the sidebar. It uses an icon instead of the text "Settings". One can choose to "Show button text labels" through Preferences.

The template title icon is not clickable. Title is not clickable. The chevron is clickable to open the drop down. I am trying to look at the clickable behavior here compared to the View - Save - Sidebar icon - Style - 3 dot menu as well as the clickable icons on the left side of the top area. The middle template title area just has a different behavior.

Clicking a template part we see this:

![Screenshot 2023-04-21 at 11 52 18](https://user-images.githubusercontent.com/5323259/233606284-76a5dafe-2f59-4c25-8980-86151bfc6f36.jpg)

Icon page icon Header chevron.

I selected a longer template name. This one looks like this:
![Screenshot 2023-04-21 at 11 54 53](https://user-images.githubusercontent.com/5323259/233606774-f0ecac25-b875-41c4-ae08-5fb0944ac193.jpg)

Going through I am hesitant in the need to add the template icon into the top bar as it almost feels a bit crowded as in too much having the icon Blog (Alternative) icon Header chevron.

As we are in the Site Editor template area as we have selected a template to view/edit. So it is obvious that we are viewing a template. Do we need the top bar template icon to give us the additional hint? 
There is no icon beside View or Save buttons.

Btw
I just noticed this area:
![Screenshot 2023-04-21 at 12 05 59](https://user-images.githubusercontent.com/5323259/233609381-c9a1783e-46e2-432d-89b8-cbbc768fed16.jpg)

Page icon will be seen in the Template sidebar area as well as in the top bar area. It feels a bit repetitive when the sidebar is open. But then again clicking the Template part we currently in Gutenberg see the same thing.

--------
Continuing....
Looking at template parts.

**Before:**
![Screenshot 2023-04-21 at 12 04 30](https://user-images.githubusercontent.com/5323259/233609115-3ac21ec0-f4df-4a26-a9f0-5aa444267b02.jpg)


**After:**
![Screenshot 2023-04-21 at 12 04 39](https://user-images.githubusercontent.com/5323259/233609144-20dc0adb-4860-40c9-8429-4696bf092a6c.jpg)

------

I believe the question I remain with is should we see the page icon - page when a part is selected.
![Screenshot 2023-04-21 at 12 12 35](https://user-images.githubusercontent.com/5323259/233610850-5b4dd8da-09ec-486b-9e12-fb29afc46c10.jpg)

Or should the page icon be hidden when a template part is selected? To bring greater focus to the template part.
![Template part only icon](https://user-images.githubusercontent.com/5323259/233611248-5335fcad-4ed7-4417-bdb0-083c9d03bac7.jpg)

Ok. I believe my rambling/brainstorming is finished.

richtabor commented:
> I believe the question I remain with is should we see the page icon - page when a part is selected.

I'm not sure it makes sense to hide it when a part is selected, other than just too much noise. But I'm still finding it difficult to justify it hiding. 

And this brings me back to the caret breadcrumb mock-ups in the header. Missing the context it provides as a breadcrumb.

jameskoster commented:
I noticed this is also included in https://github.com/WordPress/gutenberg/pull/49980, which validates the idea, but since that PR includes additional functionality around this element perhaps it makes sense to concentrate on that one?

dsas commented:
> I noticed this is also included in #49980, which validates the idea, but since that PR includes additional functionality around this element perhaps it makes sense to concentrate on that one?

The functionality in #49980 looks specific to when editing a page. The functionality in this diff would still continue to work and be useful. 

This [part of the 49980 diff](https://github.com/WordPress/gutenberg/pull/49980/files#diff-0f64190407e5a202cfef71e90e6065e3c7f6ad4fcd8601e9f3f85df137b11a92L226-R232) determines which header to use - the existing Stylebook header, the new to 49980 PageContentBreadcrumbs header or DocumentActions - the header I've modified in this diff.

We should align on using the same icon though - #49980 uses the sidebar icon rather than getting the icon from the template being edited.

jameskoster commented:
Thanks for digging into that. Agree on making the icon usage consistent – I'd lean towards using `layout` for all templates.

dsas commented:
@jameskoster this PR already does use `layout` for templates, and for template-parts the appropriate template-part icon.

Or rather, it gets the icon from the entities being edited, thus ensuring it remains consistent with everything else.

I've [commented on #49980](https://github.com/WordPress/gutenberg/pull/49980/files#r1197827834) about being consistent.

Can you help me out with next steps?

jameskoster commented:
@dsas I think this is now superseded by https://github.com/WordPress/gutenberg/pull/50369.

dsas commented:
@jameskoster that looks exciting! I'll close this issue then. Thanks for your help, and thank  you to everyone else for commenting too.
